Userland/Libraries/LibDisassembly/riscv64/Encoding.cpp

```diff
@@ -0,0 +1,549 @@
+/*
+ * Copyright (c) 2023, kleines Filmröllchen <filmroellchen@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "Encoding.h"
+#include <AK/Assertions.h>
+#include <AK/BitCast.h>
+
+namespace Disassembly::RISCV64 {
+
+// rd is always in bit positions [11:7].
+static Register extract_destination_register(u32 instruction)
+{
+    auto register_bits = (instruction >> 7) & 0b11111;
+    return static_cast<Register>(register_bits);
+}
+
+// rs1 is always in bit positions [19:15].
+static Register extract_source_register_1(u32 instruction)
+{
+    auto register_bits = (instruction >> 15) & 0b11111;
+    return static_cast<Register>(register_bits);
+}
+
+// rs2 is always in bit positions [24:20].
+static Register extract_source_register_2(u32 instruction)
+{
+    auto register_bits = (instruction >> 20) & 0b11111;
+    return static_cast<Register>(register_bits);
+}
+
+static Register extract_compressed_rs2(u16 instruction)
+{
+    auto register_bits = (instruction >> 2) & 0b11111;
+    return static_cast<Register>(register_bits);
+}
+
+// The following register formats from C only use 3 bits and offset the register number by 8:
+
+// Used as rd/rs1 in CA.
+[[maybe_unused]] static Register extract_compressed_short_rs1(u16 instruction)
+{
+    auto register_bits = (instruction >> 7) & 0b111;
+    return static_cast<Register>(register_bits + 8);
+}
+
+// Used as rs2 in CS and CA, but as rd in CL and CIW.
+[[maybe_unused]] static Register extract_compressed_short_rs2(u16 instruction)
+{
+    auto register_bits = (instruction >> 2) & 0b111;
+    return static_cast<Register>(register_bits + 8);
+}
+
+static u8 extract_funct3(u32 instruction)
+{
+    return static_cast<u8>((instruction >> 12) & 0b111);
+}
+
+static u8 extract_funct7(u32 instruction)
+{
+    return static_cast<u8>((instruction >> 25) & 0b1111111);
+}
+
+// Only used for CB format to detect C.ANDI.
+[[maybe_unused]] static u8 extract_compressed_funct2_cb(u16 instruction)
+{
+    return static_cast<u8>((instruction >> 10) & 0b11);
+}
+
+[[maybe_unused]] static u8 extract_compressed_funct2(u16 instruction)
+{
+    return static_cast<u8>((instruction >> 5) & 0b11);
+}
+
+[[maybe_unused]] static u8 extract_compressed_funct4(u16 instruction)
+{
+    return static_cast<u8>((instruction >> 12) & 0b1111);
+}
+
+[[maybe_unused]] static u8 extract_compressed_funct6(u16 instruction)
+{
+    return static_cast<u8>((instruction >> 10) & 0b111111);
+}
```

Comment:
My heart screams for a giant union/bit-field, but that might be more hassle than worth at this point,
so io

Reply:
Looking at the `.h` and the other extraction functions actually having a bunch of u32 sized structs and bit fields doesnt sound so bad

Reply:
especially as the union would be self tagging

Reply:
I really don’t want to rely on bit fields and their badly defined behavior around endianness. I think I tried doing part of this with bitfields at some point? Anyways, for many more complex encodings, bitfields make no sense at all – you need to shuffle around tons of bits. The indices here are 1-1 corresponding to the specification, making verifying this code easier.

Reply:
(If this is not possible with the DistinctNumeric, you can cheat that with an empty enum)